In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
data = pd.read_csv("./du lieu cung cap/asian_indian_recipes.csv")
data.head()

Unnamed: 0 cuisine  almond  angelica  anise  anise_seed  apple  \
0          65  indian       0         0      0           0      0   
1          66  indian       1         0      0           0      0   
2          67  indian       0         0      0           0      0   
3          68  indian       0         0      0           0      0   
4          69  indian       0         0      0           0      0   

   apple_brandy  apricot  armagnac  ...  whiskey  white_bread  white_wine  \
0             0        0         0  ...        0            0           0   
1             0        0         0  ...        0            0           0   
2             0        0         0  ...        0            0           0   
3             0        0         0  ...        0            0           0   
4             0        0         0  ...        0            0           0   

   whole_grain_wheat_flour  wine  wood  yam  yeast  yogurt  zucchini  
0                        0     0     0    0      0       0         0  
1                        0     0     0    0      0       0         0  
2                        0     0     0    0      0       0         0  
3                        0     0     0    0      0       0         0  
4                        0     0     0    0      0       1         0  

[5 rows x 385 columns]

In [3]:
data.tail()

Unnamed: 0   cuisine  almond  angelica  anise  anise_seed  apple  \
2443       57686  japanese       0         0      0           0      0   
2444       57687  japanese       0         0      0           0      0   
2445       57688  japanese       0         0      0           0      0   
2446       57689  japanese       0         0      0           0      0   
2447       57690  japanese       0         0      0           0      0   

      apple_brandy  apricot  armagnac  ...  whiskey  white_bread  white_wine  \
2443             0        0         0  ...        0            0           0   
2444             0        0         0  ...        0            0           0   
2445             0        0         0  ...        0            0           0   
2446             0        0         0  ...        0            0           0   
2447             0        0         0  ...        0            0           0   

      whole_grain_wheat_flour  wine  wood  yam  yeast  yogurt  zucchini  
2443                        0     0     0    0      0       0         0  
2444                        0     0     0    0      0       0         0  
2445                        0     0     0    0      0       0         0  
2446                        0     0     0    0      0       0         0  
2447                        0     0     0    0      0       0         0  

[5 rows x 385 columns]

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2448 entries, 0 to 2447
Columns: 385 entries, Unnamed: 0 to zucchini
dtypes: int64(384), object(1)
memory usage: 7.2+ MB


In [5]:
data.groupby(by='cuisine').count()

Unnamed: 0  almond  angelica  anise  anise_seed  apple  \
cuisine                                                            
chinese          442     442       442    442         442    442   
indian           598     598       598    598         598    598   
japanese         320     320       320    320         320    320   
korean           799     799       799    799         799    799   
thai             289     289       289    289         289    289   

          apple_brandy  apricot  armagnac  artemisia  ...  whiskey  \
cuisine                                               ...            
chinese            442      442       442        442  ...      442   
indian             598      598       598        598  ...      598   
japanese           320      320       320        320  ...      320   
korean             799      799       799        799  ...      799   
thai               289      289       289        289  ...      289   

          white_bread  white_wine  whole_grain_wheat_flour  wine  wood  yam  \
cuisine                                                                       
chinese           442         442                      442   442   442  442   
indian            598         598                      598   598   598  598   
japanese          320         320                      320   320   320  320   
korean            799         799                      799   799   799  799   
thai              289         289                      289   289   289  289   

          yeast  yogurt  zucchini  
cuisine                            
chinese     442     442       442  
indian      598     598       598  
japanese    320     320       320  
korean      799     799       799  
thai        289     289       289  

[5 rows x 384 columns]

Since the dataset is multidimensional and each attribute's presence is demonstated by 0 and 1 

--> Choose Naive Bayes algorithm to build a prediction model.

In [6]:
from sklearn.decomposition import PCA
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import BernoulliNB
from sklearn.metrics import accuracy_score
from sklearn.utils.validation import column_or_1d

In [7]:
X = data.iloc[:,2:]
y = data[['cuisine']]

In [8]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [9]:
model = BernoulliNB()

In [10]:
model.fit(X_train,column_or_1d(y_train))

BernoulliNB(alpha=1.0, binarize=0.0, class_prior=None, fit_prior=True)

In [11]:
y_pred = model.predict(X_test)

In [12]:
print('The accuracy score is: ', accuracy_score(y_test,y_pred))

The accuracy score is:  0.8285714285714286


High accuracy score with > 80% 

In [13]:
print('The train score is: ', model.score(X_train,y_train))
print('The test score is: ', model.score(X_test,y_test))

The train score is:  0.8161389172625128
The test score is:  0.8285714285714286


The model fits to 83% of the dataset and shows no overfitting (train score ~ test score). 

In [14]:
df = pd.DataFrame(y_test.iloc[:,0].values, columns=['Actual'])
df.head()

Actual
0  korean
1  korean
2  korean
3  korean
4    thai

In [15]:
df['Predicted'] = y_pred

In [16]:
df.head()

Actual Predicted
0  korean    korean
1  korean    korean
2  korean    korean
3  korean    korean
4    thai      thai

In [17]:
df.tail()

Actual Predicted
485      thai      thai
486    indian    indian
487   chinese   chinese
488    korean    korean
489  japanese   chinese

## Assessment of model

In [18]:
from sklearn.metrics import classification_report, confusion_matrix

In [19]:
print(confusion_matrix(y_test,y_pred))
print(classification_report(y_test,y_pred))

[[ 77   4   5  13   1]
 [  0 104   4   2   2]
 [  5   1  36  12   4]
 [  7   1  13 141   0]
 [  2   8   0   0  48]]
              precision    recall  f1-score   support

     chinese       0.85      0.77      0.81       100
      indian       0.88      0.93      0.90       112
    japanese       0.62      0.62      0.62        58
      korean       0.84      0.87      0.85       162
        thai       0.87      0.83      0.85        58

    accuracy                           0.83       490
   macro avg       0.81      0.80      0.81       490
weighted avg       0.83      0.83      0.83       490



Precision of japanese dishes is the lowest. They were categorized wrongly into other cuisines.  

The overall accuracy is 83%.

### Try applying PCA to simplify the model.

In [20]:
pca = PCA(n_components=3)

In [21]:
pca.fit(X)

PCA(copy=True, iterated_power='auto', n_components=3, random_state=None,
    svd_solver='auto', tol=0.0, whiten=False)

In [22]:
X_trans = pca.transform(X)

In [23]:
X_trans.shape

(2448, 3)

In [24]:
X.shape

(2448, 383)

In [25]:
X_new = pd.DataFrame(X_trans, columns=['Principal Component 1', 'Principal Component 2','Principal Component 3'])

In [26]:
X_new.head()

Principal Component 1  Principal Component 2  Principal Component 3
0               1.818150               0.825986              -0.899743
1               0.124481              -1.285016              -0.422643
2               0.045288              -1.153781              -0.439490
3               1.101262               0.659145              -0.423336
4               0.806139              -0.404010              -0.534264

In [27]:
X_train_n, X_test_n, y_train_n, y_test_n = train_test_split(X_new, y, test_size=0.2, random_state=42)

In [28]:
model_n = BernoulliNB()

In [29]:
model_n.fit(X_train_n,column_or_1d(y_train_n))

BernoulliNB(alpha=1.0, binarize=0.0, class_prior=None, fit_prior=True)

In [30]:
y_pred_n = model_n.predict(X_test_n)

In [31]:
print('The accuracy score is: ', accuracy_score(y_test_n,y_pred_n))

The accuracy score is:  0.5265306122448979


In [32]:
print('The train score is: ', model_n.score(X_train_n,y_train_n))
print('The test score is: ', model_n.score(X_test_n,y_test_n))

The train score is:  0.5474974463738509
The test score is:  0.5265306122448979


PCA does not help with predicting cuisine in the model. 

### Applying Adaboost 

In [33]:
from sklearn.ensemble import AdaBoostClassifier
from sklearn.tree import DecisionTreeClassifier

In [34]:
ml = DecisionTreeClassifier()
boosting = AdaBoostClassifier(n_estimators=100, base_estimator=ml, learning_rate=1)

In [59]:
model_new = boosting.fit(X_train,column_or_1d(y_train))

In [36]:
y_pred_b = model_new.predict(X_test)

In [37]:
print('The accuracy score is: ', accuracy_score(y_test,y_pred_b))

The accuracy score is:  0.7918367346938775


In [38]:
print('The train score is: ', model_new.score(X_train,y_train))
print('The test score is: ', model_new.score(X_test,y_test))

The train score is:  0.9964249233912156
The test score is:  0.7918367346938775


The train score is 20% higher than the test score --> overfitting.

In [39]:
print(confusion_matrix(y_test,y_pred_b))
print(classification_report(y_test,y_pred_b))

[[ 72   7   4  15   2]
 [  2 107   0   1   2]
 [  8   2  31  15   2]
 [  9   4   8 138   3]
 [  6   9   1   2  40]]
              precision    recall  f1-score   support

     chinese       0.74      0.72      0.73       100
      indian       0.83      0.96      0.89       112
    japanese       0.70      0.53      0.61        58
      korean       0.81      0.85      0.83       162
        thai       0.82      0.69      0.75        58

    accuracy                           0.79       490
   macro avg       0.78      0.75      0.76       490
weighted avg       0.79      0.79      0.79       490



### Random Forest 

In [40]:
from sklearn.ensemble import RandomForestClassifier
ml_1 = RandomForestClassifier(n_estimators=100)

In [41]:
ml_1.fit(X_train,column_or_1d(y_train))

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=100,
                       n_jobs=None, oob_score=False, random_state=None,
                       verbose=0, warm_start=False)

In [42]:
y_pred_r = ml_1.predict(X_test)

In [43]:
print('The accuracy score is: ', accuracy_score(y_test,y_pred_r))

The accuracy score is:  0.8163265306122449


In [44]:
print('The train score is: ', ml_1.score(X_train,y_train))
print('The test score is: ', ml_1.score(X_test,y_test))

The train score is:  0.9964249233912156
The test score is:  0.8163265306122449


In [45]:
print(confusion_matrix(y_test,y_pred_r))
print(classification_report(y_test,y_pred_r))

[[ 79   5   5  10   1]
 [  2 107   0   1   2]
 [  6   4  29  16   3]
 [  8   3   8 143   0]
 [  4  10   0   2  42]]
              precision    recall  f1-score   support

     chinese       0.80      0.79      0.79       100
      indian       0.83      0.96      0.89       112
    japanese       0.69      0.50      0.58        58
      korean       0.83      0.88      0.86       162
        thai       0.88      0.72      0.79        58

    accuracy                           0.82       490
   macro avg       0.80      0.77      0.78       490
weighted avg       0.81      0.82      0.81       490



#### “cumin”:1, “fish” :1, “yogurt” : 0 --> cuisine?

In [46]:
X_now = X[(X['cumin'] == 1) & (X['fish'] == 1) & (X['yogurt'] == 0)]

In [47]:
X_now.shape

(52, 383)

In [48]:
print("Cuisine predicted using Naive Bayes: ", model.predict(X_now))

Cuisine predicted using Naive Bayes:  ['indian' 'thai' 'thai' 'thai' 'thai' 'thai' 'thai' 'thai' 'thai' 'thai'
 'thai' 'thai' 'thai' 'thai' 'thai' 'thai' 'thai' 'thai' 'thai' 'thai'
 'thai' 'thai' 'thai' 'thai' 'thai' 'thai' 'thai' 'thai' 'thai' 'indian'
 'indian' 'thai' 'indian' 'thai' 'indian' 'thai' 'thai' 'thai' 'thai'
 'thai' 'thai' 'indian' 'thai' 'thai' 'thai' 'thai' 'thai' 'thai' 'thai'
 'thai' 'indian' 'thai']


In [49]:
print("Cuisine predicted using Adaboost with DecisionTree as base: ", model_new.predict(X_now))

Cuisine predicted using Adaboost with DecisionTree as base:  ['indian' 'indian' 'thai' 'thai' 'thai' 'thai' 'thai' 'thai' 'thai' 'thai'
 'thai' 'thai' 'thai' 'thai' 'thai' 'thai' 'thai' 'thai' 'thai' 'thai'
 'thai' 'thai' 'thai' 'thai' 'thai' 'thai' 'thai' 'japanese' 'japanese'
 'japanese' 'indian' 'korean' 'indian' 'thai' 'thai' 'thai' 'thai' 'thai'
 'thai' 'thai' 'thai' 'thai' 'thai' 'thai' 'thai' 'thai' 'thai' 'thai'
 'thai' 'thai' 'thai' 'thai']


In [50]:
print("Cuisine predicted using Random Forest: ", ml_1.predict(X_now))

Cuisine predicted using Random Forest:  ['indian' 'indian' 'thai' 'thai' 'thai' 'thai' 'thai' 'thai' 'thai' 'thai'
 'thai' 'thai' 'thai' 'thai' 'thai' 'thai' 'thai' 'thai' 'thai' 'thai'
 'thai' 'thai' 'thai' 'thai' 'thai' 'thai' 'thai' 'thai' 'japanese'
 'japanese' 'indian' 'korean' 'indian' 'thai' 'thai' 'thai' 'thai' 'thai'
 'thai' 'thai' 'thai' 'thai' 'thai' 'thai' 'thai' 'thai' 'thai' 'thai'
 'thai' 'thai' 'thai' 'thai']


The 3 models gives the results for dishes with cumin, fish and no yogurt belong to Indian or Thai cuisine.

#### “cumin”:1, “fish” :0, “soy_sauce” : 0 --> cuisine?

In [51]:
X_now1 = X[(X['cumin'] == 1) & (X['fish'] == 0) & (X['soy_sauce'] == 0)]

In [52]:
X_now1.shape

(375, 383)

In [53]:
print("Cuisine predicted using Naive Bayes: ", model.predict(X_now1))

Cuisine predicted using Naive Bayes:  ['indian' 'indian' 'indian' 'indian' 'indian' 'thai' 'indian' 'indian'
 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian'
 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian'
 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian'
 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian'
 'thai' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian'
 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian'
 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian'
 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian'
 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian'
 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian'
 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian'
 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian'
 'indian' 'indian

In [54]:
print("Cuisine predicted using Adaboost with DecisionTree as base: ", model_new.predict(X_now1))

Cuisine predicted using Adaboost with DecisionTree as base:  ['indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian'
 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian'
 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian'
 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian'
 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian'
 'thai' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian'
 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian'
 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian'
 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian'
 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian'
 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian'
 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian'
 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' '

In [55]:
print("Cuisine predicted using Random Forest: ", ml_1.predict(X_now1))

Cuisine predicted using Random Forest:  ['indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian'
 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian'
 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian'
 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian'
 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian'
 'thai' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian'
 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian'
 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian'
 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian'
 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian'
 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian'
 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian'
 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian' 'indian'
 'indian' 'in

The 3 models gives the results for dishes with cumin, no fish and no soy sauce belong to Indian cuisine.

By comparing confusion matrix and classification report of the 3 models, it can be seen that even though the accuracy of Naive Bayes model is not as high as the other 2 for the training test, its precision is more stable among classes. The only down point is its low precision of Japanese class. 

In [56]:
print("Naive Bayes: \n")
print(confusion_matrix(y_test,y_pred))
print(classification_report(y_test,y_pred))

Naive Bayes: 

[[ 77   4   5  13   1]
 [  0 104   4   2   2]
 [  5   1  36  12   4]
 [  7   1  13 141   0]
 [  2   8   0   0  48]]
              precision    recall  f1-score   support

     chinese       0.85      0.77      0.81       100
      indian       0.88      0.93      0.90       112
    japanese       0.62      0.62      0.62        58
      korean       0.84      0.87      0.85       162
        thai       0.87      0.83      0.85        58

    accuracy                           0.83       490
   macro avg       0.81      0.80      0.81       490
weighted avg       0.83      0.83      0.83       490



* a. Tỷ lệ % các công thức món ăn của Nhật Bản được dự đoán chính xác : **62%**
* b. Tỷ lệ % các công thức món ăn của Hàn Quốc bị gán sai nhãn thành Nhật Bản : 13/(7+1+13+141) = **8%**
* c. Nước nào có tỷ lệ % các công thức món ăn bị gán nhãn sai nhiều nhất : **Nhật Bản**

In [57]:
print("AdaBoost: \n")
print(confusion_matrix(y_test,y_pred_b))
print(classification_report(y_test,y_pred_b))

AdaBoost: 

[[ 72   7   4  15   2]
 [  2 107   0   1   2]
 [  8   2  31  15   2]
 [  9   4   8 138   3]
 [  6   9   1   2  40]]
              precision    recall  f1-score   support

     chinese       0.74      0.72      0.73       100
      indian       0.83      0.96      0.89       112
    japanese       0.70      0.53      0.61        58
      korean       0.81      0.85      0.83       162
        thai       0.82      0.69      0.75        58

    accuracy                           0.79       490
   macro avg       0.78      0.75      0.76       490
weighted avg       0.79      0.79      0.79       490



* a. Tỷ lệ % các công thức món ăn của Nhật Bản được dự đoán chính xác : **66%**
* b. Tỷ lệ % các công thức món ăn của Hàn Quốc bị gán sai nhãn thành Nhật Bản : 8/(10+5+8+141) = **5%**
* c. Nước nào có tỷ lệ % các công thức món ăn bị gán nhãn sai nhiều nhất : **Nhật Bản**

In [58]:
print("RandomForest: \n")
print(confusion_matrix(y_test,y_pred_r))
print(classification_report(y_test,y_pred_r))

RandomForest: 

[[ 79   5   5  10   1]
 [  2 107   0   1   2]
 [  6   4  29  16   3]
 [  8   3   8 143   0]
 [  4  10   0   2  42]]
              precision    recall  f1-score   support

     chinese       0.80      0.79      0.79       100
      indian       0.83      0.96      0.89       112
    japanese       0.69      0.50      0.58        58
      korean       0.83      0.88      0.86       162
        thai       0.88      0.72      0.79        58

    accuracy                           0.82       490
   macro avg       0.80      0.77      0.78       490
weighted avg       0.81      0.82      0.81       490



* a. Tỷ lệ % các công thức món ăn của Nhật Bản được dự đoán chính xác : **72%**
* b. Tỷ lệ % các công thức món ăn của Hàn Quốc bị gán sai nhãn thành Nhật Bản : 8/(10+3+8+141) = **5%**
* c. Nước nào có tỷ lệ % các công thức món ăn bị gán nhãn sai nhiều nhất : **Nhật Bản**